src/lsm/grid.zig

```diff
@@ -214,20 +227,67 @@ pub fn GridType(comptime Storage: type) type {
             for (grid.read_iop_tracer_slots) |slot| assert(slot == null);
             for (grid.write_iop_tracer_slots) |slot| assert(slot == null);
 
-            grid.cache.deinit(allocator);
-
             for (grid.cache_blocks) |block| allocator.free(block);
             allocator.free(grid.cache_blocks);
 
+            grid.cache_valid.deinit(allocator);
+            grid.cache.deinit(allocator);
+
             grid.* = undefined;
         }
 
+        pub fn cancel(grid: *Grid, callback: fn (*Grid) void) void {
+            assert(grid.canceling == null);
+
+            grid.canceling = .{ .callback = callback };
+
+            grid.read_pending_queue.reset();
+            grid.superblock.storage.reset_next_tick_lsm();
+            grid.superblock.storage.on_next_tick(
+                .vsr,
+                cancel_tick_callback,
+                &grid.canceling_next_tick,
+            );
+        }
+
+        fn cancel_tick_callback(next_tick: *NextTick) void {
+            const grid = @fieldParentPtr(Grid, "canceling_next_tick", next_tick);
+            if (grid.canceling == null) return;
+
+            assert(grid.read_pending_queue.empty());
+            assert(grid.superblock.storage.next_tick_queue_lsm.empty());
+
+            grid.cancel_join_callback();
+        }
+
+        fn cancel_join_callback(grid: *Grid) void {
+            assert(grid.canceling != null);
+            assert(grid.superblock.storage.next_tick_queue_lsm.empty());
+
+            if (grid.write_queue.empty() and
+                grid.write_iops.executing() == 0 and
+                grid.read_queue.empty())
```

Comment:
If we are waiting for R/W queues to drain, do we also ensure that nothing can add to these queues while we wait?

Reply:
Can we reset `read/write_queue` when `cancel()` is called, to speed up cancellation? We should only need to wait on the actual issued inflight I/O?

Reply:
> If we are waiting for R/W queues to drain, do we also ensure that nothing can add to these queues while we wait?

Yes; read/write_block assert that we aren't canceling.

> Can we reset read/write_queue when cancel() is called, to speed up cancellation? We should only need to wait on the actual issued inflight I/O?

:+1: